qcodes/interactive_widget.py

```diff
@@ -0,0 +1,436 @@
+    Tab,
+    Textarea,
+    VBox,
+)
+from ruamel.yaml import YAML
+
+import qcodes
+from qcodes.dataset import initialise_or_create_database_at
+from qcodes.dataset.data_export import get_data_by_id
+from qcodes.dataset.data_set import DataSet
+from qcodes.dataset.plotting import plot_dataset
+
+_META_DATA_KEY = "widget_notes"
+
+
+def _get_in(nested_keys: Sequence[str], dct: Dict) -> Dict:
+    """ Returns dct[i0][i1]...[iX] where [i0, i1, ..., iX]==nested_keys."""
+    return reduce(operator.getitem, nested_keys, dct)
+
+
+def button(
+    description: str,
+    button_style: Optional[str] = None,
+    on_click: Optional[Callable[[Any], None]] = None,
+    tooltip: Optional[str] = None,
+    layout_kwargs: Optional[Dict[str, Any]] = None,
+    button_kwargs: Optional[Dict[str, Any]] = None,
+) -> Button:
+    """Returns a `ipywidgets.Button`."""
+    layout_kwargs = layout_kwargs or {}
+    but = Button(
+        description=description,
+        button_style=button_style,
+        layout=Layout(
+            height=layout_kwargs.pop("height", "auto"),
+            width=layout_kwargs.pop("width", "auto"),
+            **layout_kwargs,
+        ),
+        tooltip=tooltip or description,
+        **(button_kwargs or {}),
+    )
+    if on_click is not None:
+        but.on_click(on_click)
+    return but
+
+
+def label(description: str) -> Label:
+    """Returns a `ipywidgets.Label` with text."""
+    return Label(
+        value=description, layout=Layout(height="max-content", width="auto")
+    )
+
+
+def _update_nested_dict_browser(
+    nested_keys: Sequence[str], table: Dict[Any, Any], box: Box
+) -> Callable[[Button], None]:
+    def update_box(_: Button) -> None:
+        box.children = (_nested_dict_browser(nested_keys, table, box),)
+
+    return update_box
+
+
+def _nested_dict_browser(
+    nested_keys: Sequence[str],
+    table: Dict[Any, Any],
+    box: Box,
+    max_nrows: int = 30,
+) -> GridspecLayout:
+    """Generates a `GridspecLayout` of the ``nested_keys`` in ``table`` which is
+    put inside of ``box``.
+
+    Whenever the table has less than ``max_nrows`` rows, the table is
+    displayed in 3 columns, otherwise it's 2 columns.
+    """
+
+    def _should_expand(x: Any) -> bool:
+        return isinstance(x, dict) and x != {}
+
+    col_widths = [8, 16, 30]
+    selected_table = _get_in(nested_keys, table)
+    nrows = (
+        sum(
+            len(v) if _should_expand(v) else 1 for v in selected_table.values()
+        )
+        + 1
+    )
+    ncols = 3
+
+    if nrows > max_nrows:
+        nrows = len(selected_table) + 1
+        col_widths.pop(1)
+        ncols = 2
+
+    grid = GridspecLayout(nrows, col_widths[-1])
+    update = partial(_update_nested_dict_browser, table=table, box=box)
+
+    # Header
+    title = " ► ".join(nested_keys)
+    grid[0, :-1] = button(title, "success")
+    up_click = update(nested_keys[:-1])
+    grid[0, -1] = button("↰", "info", up_click)
+
+    # Body
+
+    i = 1
```

Comment:
what is "i"? :) i'm basically just asking for more explicit names than I, k, v, k_, v_

Reply:
In 0757dd7ac I have renamed `i` -> `row_index`.

About the other variables, `k` and `v` are keys and values from an arbitrary dict and I don't think I can do much better. I think `(k, v)` are a commonly used notation for `(key, value)` pairs. However, if you want I can `k -> key` and `v -> value`?

Reply:
> I have renamed i -> row_index

great!

key/value would be better but it's fine as is :)